### PR TITLE
Feature/etl tester

### PR DESCRIPTION
**Pull request by myshkins**
pr to merge the airnow etl branch to develop

### CODE REVIEW

File: .env.dev [hunk 1]
```diff
@@ -11,7 +11,7 @@ FLASK_DEBUG=true
 SECRET_KEY=secret
 
 #api variables
-
+AIRNOW_KEY=AA5AB45D-9E64-41DE-A918-14578B9AC816
```
Comment (hucksy) [inline .env.dev]: we should add this to the git ignore file

File: _df_test.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import pandas as pd
+
+df1 = pd.DataFrame(
+    {
+        "name": ["bob", "alice", "joe", "jane"],
+        "p1": [12, 14, 15, 10],
+        "q1": [2, 4, 5, 8],
+    }
+)
+
```
Comment (hucksy) [inline _df_test.py]: prob don't want this test file included

File: _test.py [hunk 3]
```diff
@@ -1,3 +1,105 @@
...
+        ["latitude", "longitude", "unit", "agency name", "station id",
+         "full station id"],
+        axis=1
+    )
+    df.to_csv('site-data2.csv', header=True, index=False)
+    merged_df = shape_airnow_data(df)
+    merged_df = merged_df.replace({',': '-'}, regex=True)
+    merged_df.to_csv('merged-data.csv', header=True, index=False)
+
+extract_current_data()
```
Comment (hucksy) [inline _test.py]: should this test file also be removed?

File: merged-data.csv [hunk 4]
```diff
@@ -0,0 +1,1121 @@
+station_name,datetime,pm_10_conc,pm_10_AQI,pm_10_cat,pm_25_conc,pm_25_AQI,pm_25_AQI_cat
```
Comment (hucksy) [inline merged-data.csv]: maybe we should have a cleanup process to delete the csv files after they're loaded?

File: services/air_app/app/home/static/home.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+function hideDropDown(e) {
+    e.target.classList.add('hidden')
+}
+
+const dropDowns = document.querySelectorAll('.dropdown-btn')
+dropDowns.forEach((btn) => {
+    btn.addEventListener('click', toggleDropDown)
+    const dropDown = btn.nextElementSibling
+    dropDown.addEventListener('mouseleave', hideDropDown)
+})
```
Comment (hucksy) [inline services/air_app/app/home/static/home.js:19]: i don't know js...but look sick!

File: services/air_app/app/static/src/js/main.js [hunk 6]
```diff
@@ -1,10 +1,17 @@
-const homeButton = document.querySelector('#btn-home');
+console.log('hellow')
```
Comment (hucksy) [inline services/air_app/app/static/src/js/main.js]: we can remove the console.log

File: services/api/app/api/__init__.py [hunk 7]
```diff
@@ -1,12 +0,0 @@
...
-
-
-def create_app() -> FastAPI:
-    app = FastAPI()
-
-    @app.get("/")
-    def root():
-        return {"message": "Hello World"}
-
-    return app
```
Comment (hucksy) [inline services/api/app/api/__init__.py:12]: I like the app factory structure here...keep this maybe? Tiangolo has some great docs about using routers (which are kinda like flask blueprints): https://fastapi.tiangolo.com/tutorial/bigger-applications/

File: services/api/main.py [hunk 8]
```diff
@@ -1,4 +1,81 @@
...
+from sqlalchemy.orm import Session
+
+class TimePeriod(str, Enum):
+    day = "day"
+    week = "week"
+    month = "month"
+    year = "year"
+
+engine = create_engine(
+    "postgresql+psycopg2://airflow:airflow@postgres/airnow",
```
Comment (hucksy) [inline services/api/main.py]: I think the db uri should come from a .env variable

Also, fastapi has a slick way of getting db connections within functions. They use dependency injection, which is basically just a way of calling a function (ie, getting db connection) as a prerequisite to a function...tiangolo describes it better here: https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/

File: services/api/main.py [hunk 9]
```diff
@@ -1,4 +1,81 @@
...
+@app.get("/")
+def root():
+    return {"message": "Hello myshy"}
+
+def zipcode_to_latlong(zipcode: str):
+    """helper func, returns tuple of lat long"""
+    geo = pgeocode.Nominatim('us')
+    loc = geo.query_postal_code(zipcode)
+    return float(loc["latitude"]), float(loc["longitude"])
+
```
Comment (hucksy) [inline services/api/main.py]: ^ this is neat!

File: services/api/main.py [hunk 10]
```diff
@@ -1,4 +1,81 @@
...
+        stmt = text(
+            """
+            SELECT station_name, location_coord
+            FROM prod_airnow_stations
+            ORDER BY location_coord <-> ':y'::POINT
+            LIMIT 5
+            """
+        )
+        result = conn.execute(stmt, {"y": loc})
+        response = result.all()
```
Comment (hucksy) [inline services/api/main.py:46]: do we want just 1 result here? or 5?

File: services/api/main.py [hunk 11]
```diff
@@ -1,4 +1,81 @@
...
+        return response
+
+def query_airnow(period, station):
```
Comment (hucksy) [inline services/api/main.py]: what is the period input here? doesn't look like it's used

File: services/dags/airnow_etl.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+from datetime import datetime as dt, timedelta
+import os
+
+import pandas as pd
+import numpy as np
+import requests
+from airflow.decorators import dag, task
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.providers.postgres.hooks.postgres import PostgresHook
+
```
Comment (hucksy) [inline services/dags/airnow_etl.py]: in vscode you can right click in the editor, and there's an option for "sort imports"...which is nice to keep imports alphabetized and easier to read

File: services/dags/airnow_etl.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    "startDate": dt.utcnow().strftime('%Y-%m-%dT%H'),
+    "endDate": (dt.utcnow() + timedelta(hours=1)).strftime('%Y-%m-%dT%H'),
+    "parameters": "OZONE,PM25,PM10,CO,NO2,SO2",
+    "BBOX": "-167.716404,3.233406,-63.653904,70.867976",
+    "dataType": "B",
+    "format": "text/csv",
+    "verbose": "1",
+    "monitorType": "2",
+    "includerawconcentrations": "0",
+    "API_KEY": "AA5AB45D-9E64-41DE-A918-14578B9AC816",
```
Comment (hucksy) [inline services/dags/airnow_etl.py]: api key can use a .env variable

File: services/dags/sql/airnow_load_prod_stations.sql [hunk 14]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO prod_airnow_stations (station_name, agency_name, latitude, longitude, location_coord)
+    SELECT * FROM temp_airnow_stations
+    ON CONFLICT DO NOTHING
```
Comment (hucksy) [inline services/dags/sql/airnow_load_prod_stations.sql]: should we update on conflict?

File: services/dags/sql/airnow_load_to_prod_data.sql [hunk 15]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO prod_airnow_data (
+    station_name, reading_datetime, pm_10_conc, pm_10_AQI, 
+    pm_10_AQI_CAT, pm_25_conc, pm_25_AQI, pm_25_AQI_CAT
+    )
+SELECT * FROM temp_airnow_data
+ON CONFLICT DO NOTHING
```
Comment (hucksy) [inline services/dags/sql/airnow_load_to_prod_data.sql]: also here...update on conflict?

File: services/api/main.py [hunk 16]
```diff
@@ -1,4 +1,81 @@
...
+        q_period = timedelta(days=30)
+    if period is TimePeriod.year:
+        q_period = timedelta(days=365)
+
+    result = []
+    for station in stations:
+        s_name = station["station_name"]
+        s_data = query_airnow(q_period, s_name)
+        result.append(s_data)
+    return result
```
Comment (hucksy) [inline services/api/main.py]: similar to the comment above on routers, maybe we could organize this a bit better with a router. We'll be adding more endpoints that will use different routers.

File: services/api/main.py [hunk 17]
```diff
@@ -1,4 +1,81 @@
...
+engine = create_engine(
+    "postgresql+psycopg2://airflow:airflow@postgres/airnow",
+    future=True
+    )
+app = FastAPI()
+
+@app.get("/")
+def root():
+    return {"message": "Hello myshy"}
+
```
Comment (hucksy) [inline services/api/main.py]: we can also type hint the return values for all the functions here

**Review by hucksy — CHANGES_REQUESTED** [review]
That's a lot of code, good stuff dude. Lemme know if you want to chat about any of my comments here, some of them are definitely debatable.

File: services/dags/sql/airnow_load_prod_stations.sql (hunk 14 above)
Comment (myshkins) [reply services/dags/sql/airnow_load_prod_stations.sql]: select (specify column names)

**Pull request closed without merging**